tools/iree-prof-output-stdout.cc

```diff
@@ -6,34 +6,45 @@
 
 #include "tools/iree-prof-output-stdout.h"
 
+#include <algorithm>
 #include <cstdint>
 #include <iostream>
 #include <map>
 #include <vector>
 
+#include "third_party/abseil-cpp/absl/container/flat_hash_map.h"
 #include "third_party/abseil-cpp/absl/status/status.h"
+#include "third_party/abseil-cpp/absl/strings/str_cat.h"
 #include "third_party/tracy/public/common/TracyProtocol.hpp"
 #include "third_party/tracy/server/TracyWorker.hpp"
 #include "tools/iree-prof-output-utils.h"
 
 namespace iree_prof {
 namespace {
 
+bool HasSubstr(absl::string_view str, const std::vector<std::string>& substrs) {
+  return std::find_if(
+             substrs.begin(), substrs.end(),
+             [str](const std::string& s) { return str.find(s) != str.npos; })
+             != substrs.end();
+}
+
 template <typename T>
 struct Zone {
-  int16_t source_location_id;
+  const char* name;
   const T* zone;
 };
 
 template <typename T>
-std::vector<Zone<T>> GetTopZones(
+std::vector<Zone<T>> GetZonesFilteredAndSorted(
+    const tracy::Worker& worker,
     const tracy::unordered_flat_map<int16_t, T>& zones,
-    int num_zones_to_return) {
+    const std::vector<std::string>& zone_substrs) {
   std::map<int, Zone<T>> ordered_map;
   for (const auto& z : zones) {
-    ordered_map[z.second.total] = Zone<T>{z.first, &z.second};
-    if (ordered_map.size() > num_zones_to_return) {
-      ordered_map.erase(ordered_map.begin());
+    const char* zone_name = GetZoneName(worker, z.first);
+    if (HasSubstr(zone_name, zone_substrs)) {
+      ordered_map[z.second.total] = Zone<T>{zone_name, &z.second};
```

Comment:
There might be issue when there are two zones with the same time. I'd rather sort the vector on the second.

Reply:
It sorts zones based on total duration (z.second.total). The total duration includes all durations of the given zone even when multiple instances are running at then same time. Would it answer your concern? Please let me know if you mean something else.

Reply:
When two zones have a same total duration, the existing map entry will be overwritten by the other zone. The case would be rare, but possible.

Reply:
Right. Updated. PTAL.